File: src/main/java/com/example/gsrelationaldataaccess/CustomerController.java
package com.example.gsrelationaldataaccess;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping(value = "/customers")
public class CustomerController {
    @Autowired
    JdbcTemplate jdbcTemplate;

    @RequestMapping(value = "/{id}", method = RequestMethod.GET)
    public @ResponseBody
    Customer customer(@PathVariable long id) {
        return jdbcTemplate.queryForObject(
                "SELECT id, first_name, last_name FROM customers where id=?", new Object[]{id},
                (rs, rowNum) -> new Customer(rs.getLong("id"), rs.getString("first_name"),
                        rs.getString("last_name"))
        );
    }

    @RequestMapping(value = "", method = RequestMethod.GET)
    public @ResponseBody
    List<Customer> getCustomers() {
        return jdbcTemplate.query("SELECT id, first_name, last_name FROM customers",
                (rs, rowNum) -> new Customer(rs.getLong("id"), rs.getString("first_name"),
                        rs.getString("last_name")));
    }

    @RequestMapping(value = "/{id}", method = RequestMethod.DELETE)
    public @ResponseBody
    int deleteCustomer(@PathVariable long id) {
        return jdbcTemplate.update("DELETE FROM customers where id = ?", new Object[]{id});
    }

    @RequestMapping(value = "", method = RequestMethod.POST)
    public @ResponseBody
    int createCustomer(@RequestBody Customer c) {
        return jdbcTemplate.update("INSERT INTO customers(first_name, last_name) VALUES(?,?)",
                new Object[]{c.getFirstName(), c.getLastName()});
    }

    @RequestMapping(value = "/{id}", method = RequestMethod.PUT)
    public @ResponseBody
    int updateCustomer(@PathVariable long id, @RequestBody Customer c) {
        return jdbcTemplate.update("UPDATE customers Set first_name=?, last_name= ? where id=?",
                new Object[]{c.getFirstName(), c.getLastName(), id});
    }

}

File: src/main/scala/com/example/gsrelationaldataaccess/CustomerController.scala
package com.example.gsrelationaldataaccess


import java.sql.ResultSet

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.jdbc.core.{JdbcTemplate, RowMapper}
import org.springframework.web.bind.annotation._

@RestController
@RequestMapping(value = Array("/customers"))
class CustomerController {

  @Autowired
  val jdbcTemplate: JdbcTemplate = null

  @RequestMapping(value = Array(""))
  def getCustomers = {
    jdbcTemplate.queryForList("SELECT id, first_name, last_name FROM customers")
  }

  @GetMapping(value = Array("/{id}"))
  def getCustomer(@PathVariable id: AnyRef): Customer = {
    jdbcTemplate.queryForObject("SELECT id, first_name, last_name FROM customers where id=?", Array(id), CustomerMapper())
  }

  @DeleteMapping(value = Array("/{id}"))
  def deleteCustomer(@PathVariable id: AnyRef): Int = {
    return jdbcTemplate.update("DELETE FROM customers where id = ?", id)
  }

  @PostMapping
  def createCustomer(@RequestBody customer: Customer): Int = {
    return jdbcTemplate.update("INSERT INTO customers(first_name, last_name) VALUES(?,?)", customer.firstName, customer.lastName)
  }

  @PutMapping(value = Array("/{id}"))
  def updaetCustomer(@PathVariable id: AnyRef, @RequestBody customer: Customer): Int = {
    return jdbcTemplate.update("UPDATE customers Set first_name=?, last_name=? where id=?", customer.firstName, customer.lastName, id)
  }

}

case class CustomerMapper() extends RowMapper[Customer] {
  override def mapRow(rs: ResultSet, rowNum: Int): Customer = Customer(rs.getLong("id"), rs.getString("first_name"), rs.getString("last_name"))
}
